src/cloud_provider/helm.rs

```diff
@@ -0,0 +1,204 @@
+use crate::cloud_provider::helm::HelmAction::Deploy;
+use crate::cloud_provider::helm::HelmChartNamespaces::KubeSystem;
+use crate::cmd::helm::{helm_exec_uninstall_with_chart_info, helm_exec_upgrade_with_chart_info};
+use crate::cmd::kubectl::kubectl_exec_rollout_restart_deployment;
+use crate::error::{SimpleError, SimpleErrorKind};
+use std::path::Path;
+use std::{fs, thread};
+use thread::spawn;
+
+#[derive(Clone)]
+pub enum HelmAction {
+    Deploy,
+    Destroy,
+    Skip,
+}
+
+#[derive(Copy, Clone)]
+pub enum HelmChartNamespaces {
+    KubeSystem,
+    Prometheus,
+    Logging,
+    CertManager,
+    NginxIngress,
+    Qovery,
+}
+
+#[derive(Clone)]
+pub struct ChartSetValue {
+    pub key: String,
+    pub value: String,
+}
+
+#[derive(Clone)]
+pub struct ChartInfo {
+    pub name: String,
+    pub path: String,
+    pub namespace: HelmChartNamespaces,
+    pub action: HelmAction,
+    pub atomic: bool,
+    pub force_upgrade: bool,
+    pub timeout: String,
+    pub dry_run: bool,
+    pub wait: bool,
+    pub values: Vec<ChartSetValue>,
+    pub values_files: Vec<String>,
+}
+
+impl Default for ChartInfo {
+    fn default() -> ChartInfo {
+        ChartInfo {
+            name: "undefined".to_string(),
+            path: "undefined".to_string(),
+            namespace: KubeSystem,
+            action: Deploy,
+            atomic: true,
+            force_upgrade: false,
+            timeout: "300s".to_string(),
+            dry_run: false,
+            wait: true,
+            values: Vec::new(),
+            values_files: Vec::new(),
+        }
+    }
+}
+
+pub fn get_chart_namespace(namespace: HelmChartNamespaces) -> String {
+    match namespace {
+        HelmChartNamespaces::KubeSystem => "kube-system",
+        HelmChartNamespaces::Prometheus => "prometheus",
+        HelmChartNamespaces::Logging => "logging",
+        HelmChartNamespaces::CertManager => "cert-manager",
+        HelmChartNamespaces::NginxIngress => "nginx-ingress",
+        HelmChartNamespaces::Qovery => "qovery",
+    }
+    .to_string()
+}
+
+pub trait HelmChart {
+    fn check_prerequisites(&self) -> Result<(), SimpleError> {
+        let chart = self.get_chart_info();
+        for file in chart.values_files.clone() {
```

Comment:
use into_iter, or iter()

Reply:
Using iter() won't let me escape from the lambda on the first error